In [1]:
import numpy as np
import pandas as pd
import sqlite3
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
conn = sqlite3.connect('database.db')

In [3]:
df1 = pd.read_csv('ChicagoCensusData.csv').to_sql('CD', conn)
df2 = pd.read_csv('ChicagoPublicSchools.csv').to_sql('CPS',conn)
df3 = pd.read_csv('ChicagoCrimeData.csv').to_sql('CCD',conn)

In [4]:
%load_ext sql
%sql sqlite:///database.db

# Overview of ChicagoCensusData as CD

In [5]:
%sql SELECT * FROM CD LIMIT 3

 * sqlite:///database1.db
Done.


[(0, 1.0, 'Rogers Park', 7.7, 23.6, 8.7, 18.2, 27.5, 23939, 39.0),
 (1, 2.0, 'West Ridge', 7.8, 17.2, 8.8, 20.8, 38.5, 23040, 46.0),
 (2, 3.0, 'Uptown', 3.8, 24.0, 8.9, 11.8, 22.2, 35787, 20.0)]

# Overview of ChicagoPublicSchool as CPS

In [6]:
%sql SELECT * FROM CPS LIMIT 5

 * sqlite:///database1.db
Done.


[(0, 610038, 'Abraham Lincoln Elementary School', 'ES', '615 W Kemper Pl', 'Chicago', 'IL', 60614, '(773) 534-5720', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610038.pdf', 'Fullerton Elementary Network', 'NORTH-NORTHWEST SIDE COLLABORATIVE', 'No', 'Standard', 'Not on Probation', 'Level 1', 'Yes', 'Very Strong', 99.0, 'Very Strong', '99', 'Strong', 74.0, 'Strong', 66.0, 'Weak', '65', 'Strong', '70', 'Strong', '56', 'Average', '47', '96.00%', 2.0, '96.40%', '95.80%', '80.1', '43.3', '89.6', '84.9', '60.7', '62.6', '81.9', '85.2', '52', '62.4', '66.3', '77.9', 69.7, 64.4, 0.2, 0.9, 'Yellow', 'Green', '67.1', '54.5', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 813, 33, 'NDA', 1171699.458, 1915829.428, 41.92449696, -87.64452163, 7, 'LINCOLN PARK', 43, 18, '(41.92449696, -87.64452163)'),
 (1, 610281, 'Adam Clayton Powell Paideia Community Academy Elementary School', 'ES', '7511 S South Shore Dr', 'Chicago', 'IL', 60649, '(773) 535-6650', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610281.pdf', 'Skyway Elementary Network', 'SOUTH SIDE COLLABORATIVE', 'No', 'Track_E', 'Not on Probation', 'Level 1', 'No', 'Average', 54.0, 'Strong', '66', 'Strong', 74.0, 'Very Strong', 84.0, 'Weak', '63', 'Strong', '76', 'Weak', '46', 'Average', '50', '95.60%', 15.7, '95.30%', '100.00%', '62.4', '51.7', '21.9', '15.1', '29', '42.8', '38.5', '27.4', '44.8', '42.7', '14.1', '34.4', 16.8, 16.5, 0.7, 1.4, 'Green', 'Green', '17.2', '27.3', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 521, 46, 'NDA', 1196129.985, 1856209.466, 41.76032435, -87.55673627, 43, 'SOUTH SHORE', 7, 4, '(41.76032435, -87.55673627)'),
 (2, 610185, 'Adlai E Stevenson Elementary School', 'ES', '8010 S Kostner Ave', 'Chicago', 'IL', 60652, '(773) 535-2280', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610185.pdf', 'Midway Elementary Network', 'SOUTHWEST SIDE COLLABORATIVE', 'No', 'Standard', 'Not on Probation', 'Level 2', 'No', 'Strong', 61.0, 'NDA', 'NDA', 'Average', 50.0, 'Weak', 36.0, 'Weak', 'NDA', 'NDA', 'NDA', 'Average', '47', 'Weak', '41', '95.70%', 2.3, '94.70%', '98.30%', '53.7', '26.6', '38.3', '34.7', '43.7', '57.3', '48.8', '39.2', '46.8', '44', '7.5', '21.9', 18.3, 15.5, -0.9, -1.0, 'Red', 'Red', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 1324, 44, 'NDA', 1148427.165, 1851012.215, 41.74711093, -87.73170248, 70, 'ASHBURN', 13, 8, '(41.74711093, -87.73170248)'),
 (3, 609993, 'Agustin Lara Elementary Academy', 'ES', '4619 S Wolcott Ave', 'Chicago', 'IL', 60609, '(773) 535-4389', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_609993.pdf', 'Pershing Elementary Network', 'SOUTHWEST SIDE COLLABORATIVE', 'No', 'Track_E', 'Not on Probation', 'Level 1', 'No', 'Average', 56.0, 'Average', '44', 'Average', 45.0, 'Weak', 37.0, 'Weak', '65', 'Average', '48', 'Average', '53', 'Strong', '58', '95.50%', 10.4, '95.80%', '100.00%', '76.9', 'NDA', '26', '24.7', '61.8', '49.7', '39.2', '27.2', '69.7', '60.6', '9.1', '18.2', 11.1, 9.6, 0.9, 2.4, 'Green', 'Green', '42.9', '25', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 556, 42, 'NDA', 1164504.29, 1873959.199, 41.8097569, -87.6721446, 61, 'NEW CITY', 20, 9, '(41.8097569, -87.6721446)'),
 (4, 610513, 'Air Force Academy High School', 'HS', '3630 S Wells St', 'Chicago', 'IL', 60609, '(773) 535-1590', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610513.pdf', 'Southwest Side High School Network', 'SOUTHWEST SIDE COLLABORATIVE', 'NDA', 'Standard', 'Not on Probation', 'Not Enough Data', 'Yes', 'Average', 49.0, 'Strong', '60', 'Strong', 60.0, 'Average', 55.0, 'Weak', '45', 'Average', '54', 'Average', '53', 'Average', '49', '93.30%', 15.6, '96.90%', '100.00%', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', None, None, None, None, 'NDA', 'NDA', 'NDA', 'NDA', '14.6', '14.8', 'NDA', '16',

# Overview of ChicagoCrimeData as CCD

In [7]:
%sql SELECT * FROM CCD LIMIT 3

 * sqlite:///database1.db
Done.


[(0, 3512276, 'HK587712', '2004-08-28', '047XX S KEDZIE AVE', '890', 'THEFT', 'FROM BUILDING', 'SMALL RETAIL STORE', 0, 0, 911, 9, 14.0, 58.0, '6', 1155838.0, 1873050.0, 2004, 41.8074405, -87.70395585, '(41.8074405, -87.703955849)'),
 (1, 3406613, 'HK456306', '2004-06-26', '009XX N CENTRAL PARK AVE', '820', 'THEFT', '$500 AND UNDER', 'OTHER', 0, 0, 1112, 11, 27.0, 23.0, '6', 1152206.0, 1906127.0, 2004, 41.89827996, -87.71640551, '(41.898279962, -87.716405505)'),
 (2, 8002131, 'HT233595', '2011-04-04', '043XX S WABASH AVE', '820', 'THEFT', '$500 AND UNDER', 'NURSING HOME/RETIREMENT HOME', 0, 0, 221, 2, 3.0, 38.0, '6', 1177436.0, 1876313.0, 2011, 41.81593313, -87.62464213, '(41.815933131, -87.624642127)')]

# Tables in database FinalDB.db

In [8]:
%sql SELECT name AS Table_name from sqlite_master WHERE TYPE='table'

 * sqlite:///database1.db
Done.


[('CD',), ('CPS',), ('CCD',)]

# Column in CD

In [9]:
%sql SELECT * FROM PRAGMA_TABLE_INFO('CD')

 * sqlite:///database1.db
Done.


[(0, 'index', 'INTEGER', 0, None, 0),
 (1, 'COMMUNITY_AREA_NUMBER', 'REAL', 0, None, 0),
 (2, 'COMMUNITY_AREA_NAME', 'TEXT', 0, None, 0),
 (3, 'PERCENT_OF_HOUSING_CROWDED', 'REAL', 0, None, 0),
 (4, 'PERCENT_HOUSEHOLDS_BELOW_POVERTY', 'REAL', 0, None, 0),
 (5, 'PERCENT_AGED_16__UNEMPLOYED', 'REAL', 0, None, 0),
 (6, 'PERCENT_AGED_25__WITHOUT_HIGH_SCHOOL_DIPLOMA', 'REAL', 0, None, 0),
 (7, 'PERCENT_AGED_UNDER_18_OR_OVER_64', 'REAL', 0, None, 0),
 (8, 'PER_CAPITA_INCOME', 'INTEGER', 0, None, 0),
 (9, 'HARDSHIP_INDEX', 'REAL', 0, None, 0)]

# Columns in CPS

In [10]:
%sql SELECT * FROM PRAGMA_TABLE_INFO('CPS')

 * sqlite:///database1.db
Done.


[(0, 'index', 'INTEGER', 0, None, 0),
 (1, 'School_ID', 'INTEGER', 0, None, 0),
 (2, 'NAME_OF_SCHOOL', 'TEXT', 0, None, 0),
 (3, 'Elementary, Middle, or High School', 'TEXT', 0, None, 0),
 (4, 'Street_Address', 'TEXT', 0, None, 0),
 (5, 'City', 'TEXT', 0, None, 0),
 (6, 'State', 'TEXT', 0, None, 0),
 (7, 'ZIP_Code', 'INTEGER', 0, None, 0),
 (8, 'Phone_Number', 'TEXT', 0, None, 0),
 (9, 'Link', 'TEXT', 0, None, 0),
 (10, 'Network_Manager', 'TEXT', 0, None, 0),
 (11, 'Collaborative_Name', 'TEXT', 0, None, 0),
 (12, 'Adequate_Yearly_Progress_Made_', 'TEXT', 0, None, 0),
 (13, 'Track_Schedule', 'TEXT', 0, None, 0),
 (14, 'CPS_Performance_Policy_Status', 'TEXT', 0, None, 0),
 (15, 'CPS_Performance_Policy_Level', 'TEXT', 0, None, 0),
 (16, 'HEALTHY_SCHOOL_CERTIFIED', 'TEXT', 0, None, 0),
 (17, 'Safety_Icon', 'TEXT', 0, None, 0),
 (18, 'SAFETY_SCORE', 'REAL', 0, None, 0),
 (19, 'Family_Involvement_Icon', 'TEXT', 0, None, 0),
 (20, 'Family_Involvement_Score', 'TEXT', 0, None, 0),
 (21, 'Environment_Icon', 'TEXT', 0, None, 0),
 (22, 'Environment_Score', 'REAL', 0, None, 0),
 (23, 'Instruction_Icon', 'TEXT', 0, None, 0),
 (24, 'Instruction_Score', 'REAL', 0, None, 0),
 (25, 'Leaders_Icon', 'TEXT', 0, None, 0),
 (26, 'Leaders_Score', 'TEXT', 0, None, 0),
 (27, 'Teachers_Icon', 'TEXT', 0, None, 0),
 (28, 'Teachers_Score', 'TEXT', 0, None, 0),
 (29, 'Parent_Engagement_Icon', 'TEXT', 0, None, 0),
 (30, 'Parent_Engagement_Score', 'TEXT', 0, None, 0),
 (31, 'Parent_Environment_Icon', 'TEXT', 0, None, 0),
 (32, 'Parent_Environment_Score', 'TEXT', 0, None, 0),
 (33, 'AVERAGE_STUDENT_ATTENDANCE', 'TEXT', 0, None, 0),
 (34, 'Rate_of_Misconducts__per_100_students_', 'REAL', 0, None, 0),
 (35, 'Average_Teacher_Attendance', 'TEXT', 0, None, 0),
 (36, 'Individualized_Education_Program_Compliance_Rate', 'TEXT', 0, None, 0),
 (37, 'Pk_2_Literacy__', 'TEXT', 0, None, 0),
 (38, 'Pk_2_Math__', 'TEXT', 0, None, 0),
 (39, 'Gr3_5_Grade_Level_Math__', 'TEXT', 0, None, 0),
 (40, 'Gr3_5_Grade_Level_Read__', 'TEXT', 0, None, 0),
 (41, 'Gr3_5_Keep_Pace_Read__', 'TEXT', 0, None, 0),
 (42, 'Gr3_5_Keep_Pace_Math__', 'TEXT', 0, None, 0),
 (43, 'Gr6_8_Grade_Level_Math__', 'TEXT', 0, None, 0),
 (44, 'Gr6_8_Grade_Level_Read__', 'TEXT', 0, None, 0),
 (45, 'Gr6_8_Keep_Pace_Math_', 'TEXT', 0, None, 0),
 (46, 'Gr6_8_Keep_Pace_Read__', 'TEXT', 0, None, 0),
 (47, 'Gr_8_Explore_Math__', 'TEXT', 0, None, 0),
 (48, 'Gr_8_Explore_Read__', 'TEXT', 0, None, 0),
 (49, 'ISAT_Exceeding_Math__', 'REAL', 0, None, 0),
 (50, 'ISAT_Exceeding_Reading__', 'REAL', 0, None, 0),
 (51, 'ISAT_Value_Add_Math', 'REAL', 0, None, 0),
 (52, 'ISAT_Value_Add_Read', 'REAL', 0, None, 0),
 (53, 'ISAT_Value_Add_Color_Math', 'TEXT', 0, None, 0),
 (54, 'ISAT_Value_Add_Color_Read', 'TEXT', 0, None, 0),
 (55, 'Students_Taking__Algebra__', 'TEXT', 0, None, 0),
 (56, 'Students_Passing__Algebra__', 'TEXT', 0, None, 0),
 (57, '9th Grade EXPLORE (2009)', 'TEXT', 0, None, 0),
 (58, '9th Grade EXPLORE (2010)', 'TEXT', 0, None, 0),
 (59, '10th Grade PLAN (2009)', 'TEXT', 0, None, 0),
 (60, '10th Grade PLAN (2010)', 'TEXT', 0, None, 0),
 (61, 'Net_Change_EXPLORE_and_PLAN', 'TEXT', 0, None, 0),
 (62, '11th Grade Average ACT (2011)', 'TEXT', 0, None, 0),
 (63, 'Net_Change_PLAN_and_ACT', 'TEXT', 0, None, 0),
 (64, 'College_Eligibility__', 'TEXT', 0, None, 0),
 (65, 'Graduation_Rate__', 'TEXT', 0, None, 0),
 (66, 'College_Enrollment_Rate__', 'TEXT', 0, None, 0),
 (67, 'COLLEGE_ENROLLMENT', 'INTEGER', 0, None, 0),
 (68, 'General_Services_Route', 'INTEGER', 0, None, 0),
 (69, 'Freshman_on_Track_Rate__', 'TEXT', 0, None, 0),
 (70, 'X_COORDINATE', 'REAL', 0, None, 0),
 (71, 'Y_COORDINATE', 'REAL', 0, None, 0),
 (72, 'Latitude', 'REAL', 0, None, 0),
 (73, 'Longitude', 'REAL', 0, None, 0),
 (74, 'COMMUNITY_AREA_NUMBER', 'INTEGER', 0, None, 0),
 (75, 'COMMUNITY_AREA_NAME', 'TEXT', 0, None, 0),
 (76, 'Ward', 'INTEGER', 0, None, 0),
 (77, 'Police_District', 'INTEGER', 0, None, 0),
 (78, 'Location', 'TEXT', 0, None, 0)]

# Columns in CCD

In [11]:
%sql SELECT * FROM PRAGMA_TABLE_INFO('CCD')

 * sqlite:///database1.db
Done.


[(0, 'index', 'INTEGER', 0, None, 0),
 (1, 'ID', 'INTEGER', 0, None, 0),
 (2, 'CASE_NUMBER', 'TEXT', 0, None, 0),
 (3, 'DATE', 'TEXT', 0, None, 0),
 (4, 'BLOCK', 'TEXT', 0, None, 0),
 (5, 'IUCR', 'TEXT', 0, None, 0),
 (6, 'PRIMARY_TYPE', 'TEXT', 0, None, 0),
 (7, 'DESCRIPTION', 'TEXT', 0, None, 0),
 (8, 'LOCATION_DESCRIPTION', 'TEXT', 0, None, 0),
 (9, 'ARREST', 'INTEGER', 0, None, 0),
 (10, 'DOMESTIC', 'INTEGER', 0, None, 0),
 (11, 'BEAT', 'INTEGER', 0, None, 0),
 (12, 'DISTRICT', 'INTEGER', 0, None, 0),
 (13, 'WARD', 'REAL', 0, None, 0),
 (14, 'COMMUNITY_AREA_NUMBER', 'REAL', 0, None, 0),
 (15, 'FBICODE', 'TEXT', 0, None, 0),
 (16, 'X_COORDINATE', 'REAL', 0, None, 0),
 (17, 'Y_COORDINATE', 'REAL', 0, None, 0),
 (18, 'YEAR', 'INTEGER', 0, None, 0),
 (19, 'LATITUDE', 'REAL', 0, None, 0),
 (20, 'LONGITUDE', 'REAL', 0, None, 0),
 (21, 'LOCATION', 'TEXT', 0, None, 0)]

# Find the total number of crimes recorded in the CRIME table.

In [13]:
%sql SELECT COUNT(*) AS 'Total Crime Cases' FROM CCD

 * sqlite:///database1.db
Done.


[(533,)]

# List community area names and numbers with per capita income less than 11000.

In [14]:
%sql SELECT COMMUNITY_AREA_NAME, COMMUNITY_AREA_NUMBER FROM CD WHERE PER_CAPITA_INCOME<11000

 * sqlite:///database1.db
Done.


[('West Garfield Park', 26.0),
 ('South Lawndale', 30.0),
 ('Fuller Park', 37.0),
 ('Riverdale', 54.0)]

# List all case numbers for crimes involving minors?(children are not considered minors for the purposes of crime analysis)

In [15]:
%sql SELECT DISTINCT(DESCRIPTION) FROM CCD WHERE DESCRIPTION LIKE '%MINOR%'

 * sqlite:///database1.db
Done.


[('SELL/GIVE/DEL LIQUOR TO MINOR',), ('ILLEGAL CONSUMPTION BY MINOR',)]

In [16]:
%sql SELECT CASE_NUMBER, DESCRIPTION FROM CCD WHERE DESCRIPTION LIKE '%MINOR%'

 * sqlite:///database1.db
Done.


[('HL266884', 'SELL/GIVE/DEL LIQUOR TO MINOR'),
 ('HK238408', 'ILLEGAL CONSUMPTION BY MINOR')]

# List all kidnapping crimes involving a child?

In [17]:
%sql SELECT DISTINCT(PRIMARY_TYPE) FROM CCD

 * sqlite:///database1.db
Done.


[('THEFT',),
 ('BATTERY',),
 ('CRIMINAL DAMAGE',),
 ('NARCOTICS',),
 ('OTHER OFFENSE',),
 ('ASSAULT',),
 ('BURGLARY',),
 ('MOTOR VEHICLE THEFT',),
 ('DECEPTIVE PRACTICE',),
 ('ROBBERY',),
 ('CRIMINAL TRESPASS',),
 ('PROSTITUTION',),
 ('WEAPONS VIOLATION',),
 ('PUBLIC PEACE VIOLATION',),
 ('OFFENSE INVOLVING CHILDREN',),
 ('CRIM SEXUAL ASSAULT',),
 ('SEX OFFENSE',),
 ('INTERFERENCE WITH PUBLIC OFFICER',),
 ('GAMBLING',),
 ('LIQUOR LAW VIOLATION',),
 ('ARSON',),
 ('HOMICIDE',),
 ('KIDNAPPING',),
 ('INTIMIDATION',),
 ('STALKING',),
 ('OBSCENITY',),
 ('CONCEALED CARRY LICENSE VIOLATION',),
 ('PUBLIC INDECENCY',),
 ('NON-CRIMINAL',),
 ('OTHER NARCOTIC VIOLATION',),
 ('HUMAN TRAFFICKING',),
 ('NON - CRIMINAL',),
 ('RITUALISM',),
 ('NON-CRIMINAL (SUBJECT SPECIFIED)',),
 ('DOMESTIC VIOLENCE',)]

In [18]:
%sql SELECT * FROM CCD WHERE PRIMARY_TYPE = 'KIDNAPPING'  AND DESCRIPTION LIKE '%CHILD%'

 * sqlite:///database1.db
Done.


[(520, 5276766, 'HN144152', '2007-01-26', '050XX W VAN BUREN ST', '1792', 'KIDNAPPING', 'CHILD ABDUCTION/STRANGER', 'STREET', 0, 0, 1533, 15, 29.0, 25.0, '20', 1143050.0, 1897546.0, 2007, 41.87490841, -87.75024931, '(41.874908413, -87.750249307)')]

# List the kind of crimes that were recorded at schools. (No repetitions)

In [19]:
%sql SELECT DISTINCT(PRIMARY_TYPE) AS 'Type of crimes in School' FROM CCD WHERE COMMUNITY_AREA_NUMBER IN (SELECT COMMUNITY_AREA_NUMBER FROM CPS)

 * sqlite:///database1.db
Done.


[('THEFT',),
 ('BATTERY',),
 ('CRIMINAL DAMAGE',),
 ('NARCOTICS',),
 ('OTHER OFFENSE',),
 ('ASSAULT',),
 ('BURGLARY',),
 ('MOTOR VEHICLE THEFT',),
 ('DECEPTIVE PRACTICE',),
 ('ROBBERY',),
 ('CRIMINAL TRESPASS',),
 ('PROSTITUTION',),
 ('WEAPONS VIOLATION',),
 ('PUBLIC PEACE VIOLATION',),
 ('OFFENSE INVOLVING CHILDREN',),
 ('CRIM SEXUAL ASSAULT',),
 ('SEX OFFENSE',),
 ('INTERFERENCE WITH PUBLIC OFFICER',),
 ('GAMBLING',),
 ('LIQUOR LAW VIOLATION',),
 ('ARSON',),
 ('HOMICIDE',),
 ('KIDNAPPING',),
 ('INTIMIDATION',),
 ('STALKING',),
 ('OBSCENITY',),
 ('CONCEALED CARRY LICENSE VIOLATION',),
 ('PUBLIC INDECENCY',),
 ('NON-CRIMINAL',),
 ('OTHER NARCOTIC VIOLATION',),
 ('HUMAN TRAFFICKING',),
 ('NON - CRIMINAL',),
 ('RITUALISM',),
 ('NON-CRIMINAL (SUBJECT SPECIFIED)',)]

# List the type of schools along with the average safety score for each type.

In [20]:
%sql SELECT [Elementary, Middle, or High School], AVG(SAFETY_SCORE) FROM CPS GROUP BY [Elementary, Middle, or High School]

 * sqlite:///database1.db
Done.


[('ES', 49.52038369304557), ('HS', 49.62352941176471), ('MS', 48.0)]

# List 5 community areas with highest % of households below poverty line

In [21]:
%sql SELECT COMMUNITY_AREA_NUMBER, COMMUNITY_AREA_NAME, PERCENT_HOUSEHOLDS_BELOW_POVERTY FROM CD ORDER BY PERCENT_HOUSEHOLDS_BELOW_POVERTY DESC LIMIT 5

 * sqlite:///database1.db
Done.


[(54.0, 'Riverdale', 56.5),
 (37.0, 'Fuller Park', 51.2),
 (68.0, 'Englewood', 46.6),
 (29.0, 'North Lawndale', 43.1),
 (27.0, 'East Garfield Park', 42.4)]

# Which community area is most crime prone? Display the coumminty area number only.

In [22]:
%sql SELECT COMMUNITY_AREA_NUMBER, COUNT(*) AS 'COUNT' FROM CCD WHERE COMMUNITY_AREA_NUMBER != 'None' GROUP BY COMMUNITY_AREA_NUMBER ORDER BY COUNT(*) DESC \
LIMIT 1

 * sqlite:///database1.db
Done.


[(25.0, 43)]

# Use a sub-query to find the name of the community area with highest hardship index

In [23]:
%sql SELECT COMMUNITY_AREA_NAME FROM CD WHERE HARDSHIP_INDEX IN (SELECT MAX(HARDSHIP_INDEX) FROM CD)

 * sqlite:///database1.db
Done.


[('Riverdale',)]

# Use a sub-query to determine the Community Area Name with most number of crimes?

In [26]:
%sql SELECT COMMUNITY_AREA_NUMBER FROM CCD WHERE COMMUNITY_AREA_NUMBER != 'None' GROUP BY COMMUNITY_AREA_NUMBER ORDER BY COUNT(*) DESC LIMIT 1

 * sqlite:///database1.db
Done.


[(25.0,)]

In [24]:
%sql SELECT COMMUNITY_AREA_NAME FROM CD WHERE COMMUNITY_AREA_NUMBER IN (SELECT COMMUNITY_AREA_NUMBER FROM CCD WHERE COMMUNITY_AREA_NUMBER != 'None' GROUP BY COMMUNITY_AREA_NUMBER ORDER BY COUNT(*) DESC LIMIT 1)

 * sqlite:///database.db
Done.


[('Austin',)]

# Write and execute a SQL query to list the school names, community names and average attendance for communities with a hardship index of 98

In [25]:
%sql SELECT CPS.NAME_OF_SCHOOL, CPS.COMMUNITY_AREA_NAME, CPS.AVERAGE_STUDENT_ATTENDANCE, CD.HARDSHIP_INDEX	FROM CPS LEFT OUTER JOIN CD\
ON CPS.COMMUNITY_AREA_NUMBER=CD.COMMUNITY_AREA_NUMBER WHERE CD.HARDSHIP_INDEX='98'

 * sqlite:///database.db
Done.


[('George Washington Carver Military Academy High School', 'RIVERDALE', '91.60%', 98.0),
 ('George Washington Carver Primary School', 'RIVERDALE', '90.90%', 98.0),
 ('Ira F Aldridge Elementary School', 'RIVERDALE', '92.90%', 98.0),
 ('William E B Dubois Elementary School', 'RIVERDALE', '93.30%', 98.0)]

# Write and execute a SQL query to list all crimes that took place at a school. Include case number, crime type and community name.

In [26]:
%sql SELECT DISTINCT(LOCATION_DESCRIPTION) FROM CCD

 * sqlite:///database.db
Done.


[('SMALL RETAIL STORE',),
 ('OTHER',),
 ('NURSING HOME/RETIREMENT HOME',),
 ('RESIDENCE',),
 ('ALLEY',),
 ('PARKING LOT/GARAGE(NON.RESID.)',),
 ('STREET',),
 ('GROCERY FOOD STORE',),
 ('DEPARTMENT STORE',),
 ('BAR OR TAVERN',),
 ('RESIDENTIAL YARD (FRONT/BACK)',),
 ('RESIDENCE PORCH/HALLWAY',),
 ('CTA PLATFORM',),
 ('VACANT LOT/LAND',),
 ('SIDEWALK',),
 ('ATHLETIC CLUB',),
 ('AIRPORT/AIRCRAFT',),
 ('APARTMENT',),
 ('MOVIE HOUSE/THEATER',),
 ('VEHICLE NON-COMMERCIAL',),
 ('RESTAURANT',),
 ('RESIDENCE-GARAGE',),
 ('COMMERCIAL / BUSINESS OFFICE',),
 ('TAVERN/LIQUOR STORE',),
 ('CONSTRUCTION SITE',),
 ('GAS STATION',),
 ('DRUG STORE',),
 ('CTA BUS',),
 ('SCHOOL, PUBLIC, GROUNDS',),
 ('SCHOOL, PUBLIC, BUILDING',),
 ('CHURCH/SYNAGOGUE/PLACE OF WORSHIP',),
 ('OTHER COMMERCIAL TRANSPORTATION',),
 ('CHA APARTMENT',),
 ('WAREHOUSE',),
 ('POLICE FACILITY/VEH PARKING LOT',),
 ('HOSPITAL BUILDING/GROUNDS',),
 ('FACTORY/MANUFACTURING BUILDING',),
 ('CLEANING STORE',),
 ('CTA TRAIN',),
 ('BANK',),
 ('CHA PARKING LOT/GROUNDS',),
 ('LIBRARY',),
 ('CHA HALLWAY/STAIRWELL/ELEVATOR',),
 ('COLLEGE/UNIVERSITY GROUNDS',),
 ('SCHOOL, PRIVATE, BUILDING',),
 ('CONVENIENCE STORE',),
 ('HOUSE',),
 ('GOVERNMENT BUILDING/PROPERTY',)]

In [27]:
%sql SELECT CCD.CASE_NUMBER, CCD.PRIMARY_TYPE, CCD.LOCATION_DESCRIPTION, CD.COMMUNITY_AREA_NAME FROM CCD LEFT OUTER JOIN CD ON CCD.COMMUNITY_AREA_NUMBER=CD.COMMUNITY_AREA_NUMBER \
WHERE CCD.LOCATION_DESCRIPTION LIKE 'SCHOOL%'

 * sqlite:///database.db
Done.


[('HL353697', 'BATTERY', 'SCHOOL, PUBLIC, GROUNDS', 'South Shore'),
 ('HL725506', 'BATTERY', 'SCHOOL, PUBLIC, BUILDING', 'Lincoln Square'),
 ('HP716225', 'BATTERY', 'SCHOOL, PUBLIC, BUILDING', 'Douglas'),
 ('HH639427', 'BATTERY', 'SCHOOL, PUBLIC, BUILDING', 'Austin'),
 ('JA460432', 'BATTERY', 'SCHOOL, PUBLIC, GROUNDS', 'Ashburn'),
 ('HS200939', 'CRIMINAL DAMAGE', 'SCHOOL, PUBLIC, GROUNDS', 'Austin'),
 ('HK577020', 'NARCOTICS', 'SCHOOL, PUBLIC, GROUNDS', 'Rogers Park'),
 ('HS305355', 'NARCOTICS', 'SCHOOL, PUBLIC, BUILDING', 'Brighton Park'),
 ('HT315369', 'ASSAULT', 'SCHOOL, PUBLIC, GROUNDS', 'East Garfield Park'),
 ('HR585012', 'CRIMINAL TRESPASS', 'SCHOOL, PUBLIC, GROUNDS', 'Ashburn'),
 ('HH292682', 'PUBLIC PEACE VIOLATION', 'SCHOOL, PRIVATE, BUILDING', None),
 ('G635735', 'PUBLIC PEACE VIOLATION', 'SCHOOL, PUBLIC, BUILDING', None)]

# Write and execute a SQL statement to create a view showing the columns listed in the following table, with new column names as shown in the second column.

In [28]:
%sql CREATE VIEW V (School_Name, Safety_Rating, Family_Rating, Environment_Rating, Instruction_Rating, Leaders_Rating, Teachers_Rating) AS SELECT NAME_OF_SCHOOL, Safety_Icon, Family_Involvement_Icon, Environment_Icon, Instruction_Icon, Leaders_Icon, Teachers_Icon FROM CPS


 * sqlite:///database.db
Done.


[]

# Write and execute a SQL statement that returns all of the columns from the view.

In [29]:
%sql SELECT name FROM PRAGMA_TABLE_INFO('V')

 * sqlite:///database.db
Done.


[('School_Name',),
 ('Safety_Rating',),
 ('Family_Rating',),
 ('Environment_Rating',),
 ('Instruction_Rating',),
 ('Leaders_Rating',),
 ('Teachers_Rating',)]

# Write and execute a SQL statement that returns just the school name and leaders rating from the view.

In [30]:
%sql SELECT School_Name, Leaders_Rating FROM V LIMIT 5

 * sqlite:///database.db
Done.


[('Abraham Lincoln Elementary School', 'Weak'),
 ('Adam Clayton Powell Paideia Community Academy Elementary School', 'Weak'),
 ('Adlai E Stevenson Elementary School', 'Weak'),
 ('Agustin Lara Elementary Academy', 'Weak'),
 ('Air Force Academy High School', 'Weak')]

# Write the structure of a query to create or replace a stored procedure called UPDATE_LEADERS_SCORE that takes a in_School_ID parameter as an integer and a in_Leader_Score parameter as an integer.

#### ss_number = ss1.png

# Inside your stored procedure, write a SQL statement to update the Leaders_Score field in the CHICAGO_PUBLIC_SCHOOLS table for the school identified by in_School_ID to the value in the in_Leader_Score parameter.

#### ss_number = ss2.png

# Inside your stored procedure, write a SQL IF statement to update the Leaders_Icon field in the CHICAGO_PUBLIC_SCHOOLS table for the school identified by in_School_ID using the following information.

#### procedure is saved in mysql

####
DELIMITER $$
CREATE PROCEDURE UPDATE_LEADERS_SCORE(IN in_School_ID INT(10), IN in_Leader_Score INT(10))
BEGIN
	UPDATE chicago_public_schools
    SET Leaders_Score = in_Leader_Score
    WHERE School_ID = in_School_ID;
    
    IF (in_Leader_Score>=80) AND (in_Leader_Score<100) THEN
    	UPDATE chicago_public_schools
        SET Leaders_Icon = "Very strong"
        WHERE School_ID = in_School_ID;
    ELSEIF (in_Leader_Score>=60) AND (in_Leader_Score<=79) THEN
    	UPDATE chicago_public_schools
        SET Leaders_Icon = "Strong"
        WHERE School_ID = in_School_ID;
    ELSEIF (in_Leader_Score>=40) AND (in_Leader_Score<=59) THEN
    	UPDATE chicago_public_schools
        SET Leaders_Icon = "Average"
        WHERE School_ID = in_School_ID;
    ELSEIF (in_Leader_Score>=20) AND (in_Leader_Score<=39) THEN
    	UPDATE chicago_public_schools
        SET Leaders_Icon = "Weak"
        WHERE School_ID = in_School_ID;
    ELSEIF (in_Leader_Score>=0) AND (in_Leader_Score<=19) THEN
    	UPDATE chicago_public_schools
        SET Leaders_Icon = "Very Weak"
        WHERE School_ID = in_School_ID;
    END IF;
END$$
DELIMITER ;

# Write a query to call the stored procedure, passing a valid school ID and a leader score of 50, to check that the procedure works as expected.

# before: school id 610038, leader score = 65, leader icon = Weak
# after:  school id 610038, leader score = 50, leader icon = Average

#### CALL UPDATE_LEADERS_SCORE(610038, '50');

# ACID compatibility

####
BEGIN
	DECLARE EXIT HANDLER FOR SQLEXCEPTION
    BEGIN
    	ROLLBACK;
        RESIGNAL;
    END;
    START TRANSACTION;
	UPDATE chicago_public_schools
    SET Leaders_Score = in_Leader_Score
    WHERE School_ID = in_School_ID;
    
    IF (in_Leader_Score>=80) AND (in_Leader_Score<100) THEN
    	UPDATE chicago_public_schools
        SET Leaders_Icon = "Very strong"
        WHERE School_ID = in_School_ID;
    ELSEIF (in_Leader_Score>=60) AND (in_Leader_Score<=79) THEN
    	UPDATE chicago_public_schools
        SET Leaders_Icon = "Strong"
        WHERE School_ID = in_School_ID;
    ELSEIF (in_Leader_Score>=40) AND (in_Leader_Score<=59) THEN
    	UPDATE chicago_public_schools
        SET Leaders_Icon = "Average"
        WHERE School_ID = in_School_ID;
    ELSEIF (in_Leader_Score>=20) AND (in_Leader_Score<=39) THEN
    	UPDATE chicago_public_schools
        SET Leaders_Icon = "Weak"
        WHERE School_ID = in_School_ID;
    ELSEIF (in_Leader_Score>=0) AND (in_Leader_Score<=19) THEN
    	UPDATE chicago_public_schools
        SET Leaders_Icon = "Very Weak"
        WHERE School_ID = in_School_ID;
    ELSE 
    	ROLLBACK;
        RESIGNAL;
    END IF;
    COMMIT;
END